### ISSUE TITLE
Invalid memory access (signal 11) at address 0x0

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#8468]
https://play.crystal-lang.org/#/r/80h5

```cr
abstract class Event
end
 
class SubEvent < Event; end
 
class GenericEvent(T) < Event
end
 
class Dispatcher
  @events : Hash(Event.class, Array(Int32))
 
  def initialize
    @events = {GenericEvent => [] of Int32} of Event.class => Array(Int32)
 
    pp self
  end
end
 
Dispatcher.new
```

```
Invalid memory access (signal 11) at address 0x0
[0x55bfc65d0036] *CallStack::print_backtrace:Int32 +118
[0x55bfc65c19c9] __crystal_sigfault_handler +217
[0x7feb74cb1d00] ???
[0x55bfc65d8476] *Slice(T)::new:read_only<Pointer(UInt8), Int32, Bool>:Slice(UInt8) +22
[0x55bfc65d2ec5] *String#to_slice:Slice(UInt8) +37
[0x55bfc65d2e68] *String#to_s<String::Builder>:Nil +24
[0x55bfc65d7741] *String::Builder +17
[0x55bfc66293fa] *Event+ +26
[0x55bfc6630310] ???
```

Seems to be related to the generic event.  Setting the event hash key to `SubEvent` doesn't seem to cause the issue.

**Comment by Sija** [comment]
It also works when you pass the concrete type to `GenericEvent` (like `GenericEvent(Int32)`).
https://play.crystal-lang.org/#/r/80h7

**Comment by Blacksmoke16** [comment]
Reduced: https://play.crystal-lang.org/#/r/80ld

```cr
abstract class Event
end

class SubEvent < Event; end

class GenericEvent(T) < Event
end

pp({GenericEvent => [] of Int32} of Event.class => Array(Int32))
```

**Comment by asterite** [comment]
I don't think we should let `GenericEvent` cast to `Event`. Imagine then you do `new`, there's no way to create an generic type without type arguments.

What are you trying to do? (just to understand the use case)

**Comment by Blacksmoke16** [comment]
The use case is for an event dispatcher/listener library.  Currently I'm building out a `Hash(Event.class, Array(Callable))` where a `Callable` is essentially a proc created from a listener's instance method.

This hash is used to store the events and the listeners listening on that event.  An event is dispatched like `dispatcher.dispatch SomeEvent.new` or in the generic case like `dispatcher.dispatch GenericEvent(String).new "str"`.  Listeners are then invoked `@events[event.class].each &.call event`.

**Comment by asterite** [comment]
Okay. That means that `GenericEvent` can't be a listener, only `GenericEvent(String)`. That's why I'd like the compiler to make the above snippets be a compile error, saying that `GenericEvent` by itself can't be used, only `GenericEvent(T)`.

**Comment by Blacksmoke16** [comment]
Ah, after doing some more testing I think the problem is I was under the assumption that

```cr
GenericEvent(String).new("123").class # => GenericEvent
GenericEvent(Int32).new(123).class # => GenericEvent
```

in which case, using the base `GenericEvent` as a key would allow it to handle _ALL_ generic versions of `GenericEvent`.

I guess for my use-case I don't really _HAVE_ to have a generic.  I could just make it wrap an `Event`, or even just not include something like this by default and require users to define their own more concrete events as well as prevent event with generic type vars.

**Comment by Blacksmoke16** [comment]
So I recently been working on revamping this component, and after being away from this issue for a while I don't really know what I was thinking back then :sweat_smile: and am just going to close this.

How it currently works makes way more sense than compared to how I was thinking about it before given you wouldn't want a `GenericEvent(String)` to also invoke the listeners defined for `GenericEvent(Int32)` or what have you. The current behavior allows for this by making them separate entries in the listeners hash.

**Comment by Blacksmoke16** [comment]
Actually I should probably keep this open as this should produce an actual error instead of segfaulting like this.
